Fix and document tests

necto opened the pull request:
After the change in how `lsp-sonarlint` fetches the SonarLint executable, tests started to fail on CI, because the SonarLint executable is not downloaded automatically and `lsp-mode` skips it with only a warning

> The following servers support current file but do not have automatic installation: sonarlint

This change introduces an explicit SonarLint download step before running the tests.
I decided to make it an explicit step instead of integrating it into the test harness because it can take several minutes.

Additionally, this PR documents how you can run tests locally, and makes `lsp-sonarlint-display-rule-descr-test` more reliable.

necto committed:
Stabilize lsp-sonarlint-display-rule-descr-test

necto committed:
Fix CI tests: download SonarLint before testing

necto committed:
Document testing workflow

necto committed:
Bump NodeJS version to satisfy SonarJS analyzer requirements

necto commented:
There are 3 kinds of failures remaining:
1. windows-latest + emacs snapshot
  This seems to be Eask problem - it crashes on the emacs `(read-from-whole-string ARG)` function.
  It might be that Emacs or Eask is improperly installed on Windows.
  I have no Windows machine and little Windows skill to investigate.
  I did manage to hit similar bug when I was trying to install Emacs snapshot on my linux machine with 
      ```
      guix shell -K --without-tests=emacs-pgtk --with-git-url=emacs-pgtk=https://github.com/emacs-mirror/emacs --with-commit=emacs-pgtk=ce56f939affe8147f4172aa030058029a3c7922b emacs-pgtk
      ```
  The installation command failed some tests (despite `--without-tests` flag), but I was able to get the Emacs 31.0.5 binary.
  When I put it into `PATH` and tried to run `eask status`, I got the same error:
  > Can't read whole string

  Note that this job [was already failing before this change](https://github.com/emacs-lsp/lsp-sonarlint/actions/runs/9719294030/job/26828827726).
  I will remove it from CI for now, and let someone with Windows capability resuscitate it when needed.

2. MacOS/Ubuntu + Emacs snapshot
  The tests run ad infinum printing "Timeout while waiting for response.  Method: shutdown".
  In a normal run, this error is printed only a couple of times per test (that's the way we shutdown SonarLint LSP server).
  It seems something is odd with the snapshot + it does not play well with GitHub Actions.
  I was able to reproduce the crazy error-repetition locally using [act](https://github.com/nektos/act)
  However, unlike GH Action, the tests printed a lot of copies of this same error message and then succeeded in regular time.
  When I attached to the docker container produced by `act` I was able to run the tests manually
  And I've seen many copies of the same error message.
  However, when ran interactively from Emacs within that container,
  I only see normal amount of these messages in the `*Messages*` buffer
  I will remove testing Emacs snapshot versions for now.
  
 3. MacOS all Emacs versions
   Here `lsp-sonarlint-c++-reports-issues` fails (times out). I don't know what is going on here (maybe an issue with path to compiler_commands.json?). @mjburling apparently succeeded at using C++ analyzer, so I would appreciate if you, @mjburling could run the test suite (as explained in the readme version in this PR) and report if it runs well on your machine.

jcs090218 commented:
> 1. windows-latest + emacs snapshot

You can ignore this error for now, see https://github.com/emacs-eask/cli/issues/224.

necto committed:
use any c++ compiler in your path, not just /usr/bin/c++

As an example, guix package manager installs binaries in other places,
so you won't find /usr/bin/c++ on a guix-managed system.

necto committed:
Add a test checking the required C++ compiler

necto committed:
Fix regex extracting compiler exec

necto committed:
Check that LSP detect absence of SonarLint

necto committed:
Exclude platforms with failing tests

jcs090218 commented on .github/workflows/test.yml:
```diff
@@ -17,22 +17,25 @@ jobs:
+        #     experimental: true
+        #   - os: windows-latest
+        #     emacs-version: snapshot
+        #     experimental: true
```
We can just remove the `windows-latest` test.

necto replied:
removed

jcs090218 replied:
Sorry for the confusion. I mean comment out only the windows snapshot.

necto replied:
The other two are problematic as well. as you can see, they take all the time in the world (6h), even though they will eventually pass.

jcs090218 replied:
Ah, okay. I guess we can just comment them all out.

necto replied:
done 33f07a2e1c2e6a938fffe410795390ec2ee0242f

necto committed:
Remove windows from list of platforms to test on

necto committed:
Bump emacs 29.x to 29.4

jcs090218 commented on .github/workflows/test.yml:
```diff
@@ -17,22 +17,22 @@ jobs:
       matrix:
-        os: [ubuntu-latest, macos-latest]
+        # TODO: fix lsp-sonarlint-c++-reports-issues on macos and include it
+        os: [ubuntu-latest]
```
Can we ignore specific tests instead of entirely removing the `macos-latest`? 🤔

necto replied:
Indeed, that's better. I ignored just `lsp-sonarlint-c++-reports-issues` on macos

necto committed:
Re-enable MacOS except for one test

necto committed:
Simplify a test

necto committed:
Disable all snapshot tests

necto commented:
The tests fail now due to a problem with setup-emacs GH Action. I cannot rerun them to see if the problem is fixed.

jcs090218 commented:
LGTM. Thank you for your effort on the CI tests! :D
